File: tron-backend/src/adapters/pm-orchestrator.js
// src/adapters/pm-orchestrator.js
const BasecampAdapter = require('./basecamp');
const JiraAdapter = require('./jira');
const MondayAdapter = require('./monday');

class PMOrchestrator {
    
    static async getTickets(pmConfig = {}, mapping = {}, orgId) {
        const provider = pmConfig.provider || pmConfig.pm_provider || 'none';
        const projectId = pmConfig.project_id || pmConfig.pm_project_id || pmConfig.board_id;
        
        console.log(`🔍 [ORCHESTRATOR DEBUG] Received mapping:`, JSON.stringify(mapping, null, 2));
        
        try {
            if (provider === 'basecamp') {
                if (!orgId) throw new Error("Missing orgId for Basecamp request.");
                
                const actualMapping = mapping.mapping ? mapping.mapping : mapping;
                
                const todoTasks = actualMapping.todo 
                    ? await BasecampAdapter.fetchActiveTasks(projectId, actualMapping.todo, orgId) 
                    : [];
                
                const inProgressCol = actualMapping.branch_created || actualMapping.in_progress;
                
                console.log(`🔍 [ORCHESTRATOR] Fetching In-Progress tasks from column: ${inProgressCol}`);

                const inProgressTasks = inProgressCol 
                    ? await BasecampAdapter.fetchActiveTasks(projectId, inProgressCol, orgId) 
                    : [];
                
                return [
                    ...todoTasks.map(t => ({ ...t, state: 'To Do' })),
                    ...inProgressTasks.map(t => ({ ...t, state: 'In Progress' }))
                ];
            } else if (provider === 'jira') {
                return await JiraAdapter.fetchActiveTasks(pmConfig.project_key);
            } else if (provider === 'monday') { 
                return await MondayAdapter.fetchActiveTasks(projectId);
            }
        } catch (error) {
            console.error(`❌ [ORCHESTRATOR] Failed to fetch tickets for ${provider}:`, error.message);
        }
        return [];
    }

    static async updateTicketStatus(pmConfig = {}, ticketId, newStatusID, orgId) {
        const provider = pmConfig.provider || pmConfig.pm_provider || 'none';
        const projectId = pmConfig.project_id || pmConfig.pm_project_id || pmConfig.board_id;
        
        if (!newStatusID) {
            console.log(`⏭️ [ORCHESTRATOR] Skipping update: No destination column ID provided.`);
            return;
        }

        try {
            if (provider === 'basecamp') {
                if (!orgId) throw new Error("Missing orgId for Basecamp request.");
                await BasecampAdapter.updateTicketStatus(ticketId, newStatusID, projectId, orgId); 
            } else if (provider === 'jira') {
                await JiraAdapter.updateTicketStatus(ticketId, newStatusID);
            } else if (provider === 'monday') { 
                await MondayAdapter.updateTicketStatus(ticketId, newStatusID, projectId);
            }
        } catch (error) {
            console.error(`❌ [ORCHESTRATOR] Failed to update status for ${provider}:`, error.message);
            throw error; 
        }
    }

    static async resolveTask(pmConfig = {}, taskName, mapping = {}, orgId) {
        const provider = pmConfig.provider || pmConfig.pm_provider || 'none';
        const projectId = pmConfig.project_id || pmConfig.pm_project_id || pmConfig.board_id;
        const fallbackId = taskName.replace(/[^a-zA-Z0-9]/g, '-').toLowerCase();

        try {
            if (provider === 'basecamp') {
                if (!orgId) throw new Error("Missing orgId for Basecamp request.");
                
                const actualMapping = mapping.mapping ? mapping.mapping : mapping;
                return await BasecampAdapter.resolveTask(projectId, actualMapping.todo, taskName, orgId);
            } else if (provider === 'jira' || provider === 'monday') {
                return fallbackId; 
            }
        } catch (error) {
            console.error(`❌ [ORCHESTRATOR] Failed to resolve task for ${provider}:`, error.message);
        }
        
        return fallbackId;
    }

    static async assignTicket(pmConfig = {}, ticketId, developer, orgId) {
        const provider = pmConfig.provider || pmConfig.pm_provider || 'none';
        const projectId = pmConfig.project_id || pmConfig.pm_project_id || pmConfig.board_id;
        
        // 🌟 THE FIX: Removed the hardcoded 'dev' check.
        // We only abort if the developer string is completely missing or empty.
        if (!developer || String(developer).trim() === '') {
            console.log(`⏭️ [ORCHESTRATOR] Skipping assignment: No developer provided.`);
            return; 
        }

        try {
            if (provider === 'basecamp') {
                if (!orgId) throw new Error("Missing orgId for Basecamp request.");
                
                // We now hand the raw string off to the BasecampAdapter,
                // which is responsible for matching the string to an actual Basecamp Member ID.
                await BasecampAdapter.assignDeveloper(projectId, ticketId, developer, orgId); 
            }
        } catch (error) {
            console.error(`❌ [ORCHESTRATOR] Failed to assign ticket for ${provider}:`, error.message);
        }
    }
}

module.exports = PMOrchestrator;

File: tron-go-backend/internal/services/pm_orchestrator.go
package services

import (
	"fmt"
	"regexp"
	"strings"
)

// PMAdapter is the interface that breaks the import cycle.
type PMAdapter interface {
	FetchActiveTasks(projectID, columnID, orgID string) ([]map[string]interface{}, error)
	UpdateTicketStatus(exactUrl, newColumnID, projectID, orgID string) error
	ResolveTask(projectID, todoColumnID, taskName, orgID string) (string, string, error)
	AssignDeveloper(exactUrl, developerName, orgID string) error
}

type Ticket struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	State       string `json:"state"`
}

type PMOrchestrator struct {
	Basecamp PMAdapter
}

func NewPMOrchestrator(basecamp PMAdapter) *PMOrchestrator {
	return &PMOrchestrator{
		Basecamp: basecamp,
	}
}

func extractMappingValue(mapping map[string]interface{}, key string) string {
	if nested, ok := mapping["mapping"].(map[string]interface{}); ok {
		if val, exists := nested[key].(string); exists {
			return val
		}
	}
	if val, exists := mapping[key].(string); exists {
		return val
	}
	return ""
}

func (orch *PMOrchestrator) GetTickets(provider, projectID, orgID string, mapping map[string]interface{}) []Ticket {
	provider = strings.ToLower(provider)
	var allTickets []Ticket

	fmt.Printf("🔍 [ORCHESTRATOR DEBUG] Routing GetTickets to provider: %s\n", provider)

	switch provider {
	case "basecamp":
		if orgID == "" {
			fmt.Println("❌ [ORCHESTRATOR] Missing orgId for Basecamp request.")
			return allTickets
		}

		todoCol := extractMappingValue(mapping, "todo")
		inProgressCol := extractMappingValue(mapping, "branch_created")
		if inProgressCol == "" {
			inProgressCol = extractMappingValue(mapping, "in_progress")
		}

		if todoCol != "" {
			tasks, _ := orch.Basecamp.FetchActiveTasks(projectID, todoCol, orgID)
			for _, t := range tasks {
				// 🌟 THE FIX: Safely extract the ID without scientific notation!
				idStr := fmt.Sprintf("%v", t["id"])
				if fVal, ok := t["id"].(float64); ok {
					idStr = fmt.Sprintf("%.0f", fVal)
				}

				allTickets = append(allTickets, Ticket{
					ID:          idStr,
					Title:       fmt.Sprintf("%v", t["title"]),
					Description: fmt.Sprintf("%v", t["description"]),
					State:       "To Do",
				})
			}
		}

		if inProgressCol != "" {
			fmt.Printf("🔍 [ORCHESTRATOR] Fetching In-Progress tasks from column: %s\n", inProgressCol)
			tasks, _ := orch.Basecamp.FetchActiveTasks(projectID, inProgressCol, orgID)
			for _, t := range tasks {
				// 🌟 THE FIX: Apply safe extraction here as well!
				idStr := fmt.Sprintf("%v", t["id"])
				if fVal, ok := t["id"].(float64); ok {
					idStr = fmt.Sprintf("%.0f", fVal)
				}

				allTickets = append(allTickets, Ticket{
					ID:          idStr,
					Title:       fmt.Sprintf("%v", t["title"]),
					Description: fmt.Sprintf("%v", t["description"]),
					State:       "In Progress",
				})
			}
		}

	case "jira", "monday":
		// Phase 4 implementations
	}

	return allTickets
}

func (orch *PMOrchestrator) UpdateTicketStatus(provider, projectID, exactUrl, newStatusID, orgID string) error {
	provider = strings.ToLower(provider)

	if newStatusID == "" {
		fmt.Println("⏭️ [ORCHESTRATOR] Skipping update: No destination column ID provided.")
		return nil
	}

	switch provider {
	case "basecamp":
		if orgID == "" {
			return fmt.Errorf("missing orgId for Basecamp request")
		}
		return orch.Basecamp.UpdateTicketStatus(exactUrl, newStatusID, projectID, orgID)
	}

	return nil
}

func (orch *PMOrchestrator) ResolveTask(provider, projectID, taskName, orgID string, mapping map[string]interface{}) (string, string) {
	provider = strings.ToLower(provider)
	re := regexp.MustCompile(`[^a-zA-Z0-9]`)
	fallbackID := strings.ToLower(re.ReplaceAllString(taskName, "-"))

	switch provider {
	case "basecamp":
		if orgID == "" {
			fmt.Println("❌ [ORCHESTRATOR] Missing orgId for Basecamp request.")
			return fallbackID, ""
		}
		todoCol := extractMappingValue(mapping, "todo")

		taskID, exactUrl, err := orch.Basecamp.ResolveTask(projectID, todoCol, taskName, orgID)
		if err == nil && taskID != "" {
			return taskID, exactUrl
		}

	case "jira", "monday":
		return fallbackID, ""
	}

	return fallbackID, ""
}

func (orch *PMOrchestrator) AssignTicket(provider, projectID, exactUrl, developer, orgID string) {
	provider = strings.ToLower(provider)

	if strings.TrimSpace(developer) == "" {
		fmt.Println("⏭️ [ORCHESTRATOR] Skipping assignment: No developer provided.")
		return
	}

	switch provider {
	case "basecamp":
		if orgID == "" {
			fmt.Println("❌ [ORCHESTRATOR] Missing orgId for Basecamp request.")
			return
		}
		orch.Basecamp.AssignDeveloper(exactUrl, developer, orgID)
	}
}
